Dockerfile

```diff
@@ -1,50 +1,61 @@
-# Build 
+# Builder
 # --------------------------------------------------------
 
-FROM golang:1.18.2-alpine3.15 as build
+FROM golang:${GO_VERSION}-alpine as builder
 
-# linux-headers needed by github.com/zondax/hid
 RUN set -eux; apk add --no-cache ca-certificates build-base; apk add git linux-headers
 
-# CosmWasm: see https://github.com/CosmWasm/wasmvm/releases
-# 1) Add the releases into /lib/
-ADD https://github.com/CosmWasm/wasmvm/releases/download/v1.0.0/libwasmvm_muslc.aarch64.a \ 
-  https://github.com/CosmWasm/wasmvm/releases/download/v1.0.0/libwasmvm_muslc.x86_64.a \ 
-  /lib/
-# 2) Verify their hashes
-RUN sha256sum /lib/libwasmvm_muslc.aarch64.a | grep 7d2239e9f25e96d0d4daba982ce92367aacf0cbd95d2facb8442268f2b1cc1fc; \
-  sha256sum /lib/libwasmvm_muslc.x86_64.a | grep f6282df732a13dec836cda1f399dd874b1e3163504dbd9607c6af915b2740479
-# Copy the right library according to architecture. The final location will be found by the linker flag `-lwasmvm_muslc`
-RUN cp /lib/libwasmvm_muslc.$(uname -m).a /lib/libwasmvm_muslc.a
-
-# Get go dependencies first
-COPY go.mod go.sum /osmosis/
+# Download go dependencies
 WORKDIR /osmosis
+COPY go.mod go.mod
+COPY go.sum go.sum
 RUN --mount=type=cache,target=/root/.cache/go-build \
- --mount=type=cache,target=/root/go/pkg/mod \
- go mod download
+    --mount=type=cache,target=/root/go/pkg/mod \
+    go mod download
 
-# Copy all files into our docker repo
-COPY . /osmosis
+# Cosmwasm - download correct libwasmvm version
+RUN WASMVM_VERSION=$(go list -m github.com/CosmWasm/wasmvm | cut -d ' ' -f 2) && \
+    wget https://github.com/CosmWasm/wasmvm/releases/download/$WASMVM_VERSION/libwasmvm_muslc.$(uname -m).a \
+      -O /lib/libwasmvm_muslc.a
 
-# build
+# Cosmwasm - verify checksum
+RUN wget https://github.com/CosmWasm/wasmvm/releases/download/v1.0.0/checksums.txt -O /tmp/checksums.txt && \
+    sha256sum /lib/libwasmvm_muslc.a | grep $(cat /tmp/checksums.txt | grep $(uname -m) | cut -d ' ' -f 1)
 
+# Copy the remaining files
+COPY . .
+
+# Build osmosisd binary
 RUN --mount=type=cache,target=/root/.cache/go-build \
-  --mount=type=cache,target=/root/go/pkg/mod \
-  BUILD_TAGS=muslc LINK_STATICALLY=true make build
+    --mount=type=cache,target=/root/go/pkg/mod \
+    VERSION=$(echo $(git describe --tags) | sed 's/^v//') && \
+    COMMIT=$(git log -1 --format='%H') && \
+    go build \
+      -mod=readonly \
+      -tags "netgo,ledger,muslc" \
+      -ldflags "-X github.com/cosmos/cosmos-sdk/version.Name="osmosis" \
+              -X github.com/cosmos/cosmos-sdk/version.AppName="osmosisd" \
+              -X github.com/cosmos/cosmos-sdk/version.Version=$VERSION \
+              -X github.com/cosmos/cosmos-sdk/version.Commit=$COMMIT \
+              -X github.com/cosmos/cosmos-sdk/version.BuildTags='netgo,ledger,muslc' \
+              -w -s -linkmode=external -extldflags '-Wl,-z,muldefs -static'" \
+      -trimpath \
+      -o /osmosis/build/ \
+      ./...
```

Comment:
Is there a make command we should be running, instead of copying this here?

Reply:
Not sure if there is already a `make` target that will generate the `go build` command with all these flags.

We could for sure write one but avoiding using the Makefile in the Dockerfile was intentional. It makes the image creation more transparent and immediate.